### PR TITLE
Adding taquito plugin tests to verify that contract(s) can be deployed.

### PR DESCRIPTION
**Pull request by sinapsist**
Adding tests to verify possible scenarios for the taquito plugin including both positive and negative scenarios.

### CODE REVIEW

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import {generateTestProject} from "./utils/utils";
+import fs from "fs";
+import {execSync} from "child_process";
+import path from "path";
+import waitForExpect from "wait-for-expect";
+import { TezosToolkit } from '@taquito/taquito';
+
+const tezos = new TezosToolkit('https://hangzhounet.api.tez.ie');
```
Comment (alexzbusko) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:8]: Since there is only 1 describe block I would add these things below there to break up things getting imported and things getting initialized

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+            await waitForExpect(() => {
+
+                // 2. Run taq deploy on a selected test network described in "test" environment
+                const [,,,stdoutDeploy] = execSync(`taq deploy -e ${environment}`, {cwd: `./${taqueriaProjectPath}`}).toString().trim().split(/\r?\n/);
+
+                // 3. Verify that contract has been originated on the network
+                expect(stdoutDeploy).toContain("hello-tacos.tz");
+                expect(stdoutDeploy).toContain("hangzhounet");
+                const [,,contractHash] = stdoutDeploy.split("│");
+                smartContractHash = contractHash.trim();
```
Comment (alexzbusko) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:42]: You're setting smartContractHash and then not using it in the next line for expecting

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        }
+
+    });
+
+    // Remove all files from artifacts folder without removing folder itself
+    afterEach(() => {
+        try {
+            const files = fs.readdirSync(`${taqueriaProjectPath}/artifacts/`);
+            for (const file of files) {
+                fs.unlinkSync(path.join(`${taqueriaProjectPath}/artifacts/`, file));
```
Comment (alexzbusko) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:328]: Instead of unlink can you use fs.rmSync to remove the files?

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+                expect(contract.address).toBe(smartContractHash);
+            });
+
+        } catch(error) {
+            throw new Error (`error: ${error}`);
+        }
+
+    });
+
+    test('Verify that taqueria taquito plugin can deploy one contract using originate {contractName} command', async () => {
```
Comment (alexzbusko) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:97]: Since this test is only testing that the originate alias is mapped to the deploy command I think you can get the same verification that originate will work by not deploying a contract but comparing the output of the `taq deploy --help` and `taq originate --help` and making sure they are the same. The previous test verifies that the contract can be deployed/originated and this one shouldn't have to test that again

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 2 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:42]: Got point. Thank you! Fixed.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 1 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:8]: Sure. I can do it. Fixed.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 3 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:328]: No, because rmSync also deletes the folder, that causes an issue for taqueria for the next test, because it cannot find the "artifacts" folder.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 4 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:97]: Let's ask @mweichert I am just afraid that testing help is not a functional test to cover it. I will post the question into discussion.

File: tests/e2e/data/config-taquito-test-environment-invalid-config-network-url.json [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            "protocol": "PtHangz2aRngywmSRGGvrcTyMbbdpWdpFKuS4uMWxg2RaH9i1qx",
+            "rpcUrl": "http://localhost:20000"
+        }
+    },
+    "network": {
+        "hangzhounet": {
+            "label": "hangzhounet",
+            "rpcUrl": "https://invalid.test",
+            "protocol": "PtHangz2aRngywmSRGGvrcTyMbbdpWdpFKuS4uMWxg2RaH9i1qx",
+            "faucet": {
```
Comment (mweichert) [inline tests/e2e/data/config-taquito-test-environment-invalid-config-network-url.json:57]: Ivan, one thing we have to consider - eventually after running these tests enough times, our account balances will be depleted. 

Do you mind creating an issue to look into this and add the link as a todo comment here please?

I believe we can use the keygen service to get faucets programmatically which should resolve this:
https://docs.google.com/presentation/d/1Oe6ZOE7_mX8bDh-qNKLBrsNFJzOOiELbwbz5AzOgtYs/edit#slide=id.gf347d25493_0_0

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+            // More details about the error - "counter %x already used for contract %y"
+            // It is applicable only for auto-tests that run very fast
+            // Uses retry mechanism to avoid test to fail
+
+            await waitForExpect(() => {
+
+                // 2. Run taq deploy on a selected test network described in "test" environment
+                const [,,,stdoutDeploy] = execSync(`taq deploy -e ${environment}`, {cwd: `./${taqueriaProjectPath}`}).toString().trim().split(/\r?\n/);
+
+                // 3. Verify that contract has been originated on the network
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:39]: Ivan, unfortunately checking the name here is probably not enough to assure that a contract has been originated properly.

We should grab the contract address and use curl or taquito to make a rpc call that verifies whether the address is valid.

To get the contract address, you can see how I did that in the hello-tacos dapp here: https://github.com/ecadlabs/hello-tacos/blob/taqified/taqification-steps.md#environments

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                expect(stdoutDeploy).toContain("hello-tacos.tz");
+                expect(stdoutDeploy).toContain("hangzhounet");
+                const [,,contractHash] = stdoutDeploy.split("│");
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:42]: The commas look like syntax typos here.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 6 above)
Comment (mweichert) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:39]: I commented too early - you did exactly what I was asking later on in the test case. :)

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+            await waitForExpect(async () => {
+                const contract = await tezos.contract.at(smartContractHash);
+                expect(contract.address).toBe(smartContractHash);
+            });
+
+        } catch(error) {
+            throw new Error (`error: ${error}`);
+        }
+
+    });
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:57]: We should add an additional step here that verifies that no other contracts were originated. That is, because 'taq originate [contract]` and `taq origiante` are both valid, we're not testing whether the first is properly interpreted.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 4 above)
Comment (mweichert) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:98]: I wouldn't have caught this. Good thoughts Alex.

As the tests probably take a long time to run, it probably does make sense to optimize things to take the least amount of time as possible.

Nonetheless, given that this works, I'd probably suggest we let this go for now, create an issue to come back to this, and add a todo comment with a link to the issue.

In the issue, we can list a few options we're considering:
- Comparing the output of `taq deploy --help` and `taq originate --help`
- Grepping `taq deploy --help`
- Parsing the `state.json` file
- Parsing the pluginInfo response

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+            // 4. Verify that contracts have been originated to the network
+            await waitForExpect(async () => {
+                const contractOne = await tezos.contract.at(smartContractOneHash);
+                expect(contractOne.address).toBe(smartContractOneHash);
+                const contractTwo = await tezos.contract.at(smartContractTwoHash);
+                expect(contractTwo.address).toBe(smartContractTwoHash);
+            });
+
+        } catch(error) {
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:177]: Ivan, can we also verify please that the contract addresses are unique?

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+            throw new Error (`error: ${error}`);
+        }
+
+    });
+
+    // TODO: Currently there is no specific message for this type of error
+    // Issue associated https://github.com/ecadlabs/taqueria/issues/313
+    test.skip('Verify that taqueria taquito plugin will show proper error when environment does not exists', async () => {
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:185]: Ivan, do you mind creating an issue to restore this test case once #313 is resolved, and add the link to that issue as a TODO comment here. In the issue description, you can note that the issue is blocked by #313 which will link the two issues together.

Comment (mweichert) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:185]: Btw, I think this particular problem is resolved in the upcoming v0.0.9 release.

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+            // 3. Verify that proper error displays in the console
+            // expect(stdoutDeploy).toContain("E_INVALID_PLUGIN_RESPONSE");
+
+        } catch(error) {
+            throw new Error (`error: ${error}`);
+        }
+
+    });
+
+    // TODO: Currently there is no specific message for this type of error
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:207]: Please see comment above. We can probably create one issue for enabling any of these skipped tests that are dependent on this problem being resolved.

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+            expect(stdoutDeploy).toContain("E_INVALID_PLUGIN_RESPONSE");
+            expect(stdoutDeploy).toContain('\\"status\\":\\"failed\\"');
+
+        } catch(error) {
+            throw new Error (`error: ${error}`);
+        }
+
+    });
+
+    // TODO: Currently there is no specific message for this type of error
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:254]: Please see comment above. We can probably create one issue for enabling any of these skipped tests that are dependent on this problem being resolved.

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+            // 3. Verify that proper error displays in the console
+            expect(stdoutDeploy).toContain("E_INVALID_PLUGIN_RESPONSE");
+
+        } catch(error) {
+            throw new Error (`error: ${error}`);
+        }
+
+    });
+
+    // TODO: Currently there is no specific message for this type of error
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:277]: Please see comment above. We can probably create one issue for enabling any of these skipped tests that are dependent on this problem being resolved.

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
```
Comment (mweichert) [inline tests/e2e/taqueria-plugin-taquito.spec.ts:300]: Please see comment above. We can probably create one issue for enabling any of these skipped tests that are dependent on this problem being resolved.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 7 above)
Comment (alexzbusko) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:42]: Split returns an array so if only 1 part is needed you can just take the index of the one you know it should be

**Review by mweichert — CHANGES_REQUESTED** [review]
A few items here, but most are related to created issues for work left to do or that is blocked due to another issue.

Note, I'd like it if you could come back to this after v0.0.9 is released as I believe it addresses the problem with errors not being emitted for invalid environment and network configurations.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 6 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:39]: :)

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 10 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:185]: I checked 0.1.1 and the issue is still there. 
Greenshot 2022-02-24 07 41 49

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 14 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:300]: Fixed. Issue to investigate and re-enable tests - https://github.com/ecadlabs/taqueria/issues/377

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 13 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:277]: Fixed. Issue to investigate and re-enable tests - https://github.com/ecadlabs/taqueria/issues/377

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 12 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:254]: Fixed. Issue to investigate and re-enable tests - https://github.com/ecadlabs/taqueria/issues/377

File: tests/e2e/taqueria-plugin-taquito.spec.ts [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
```
[the remark this answers is not in the document]
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:230]: Fixed. Issue to investigate and re-enable tests - https://github.com/ecadlabs/taqueria/issues/377

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 11 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:207]: Fixed. Issue to investigate and re-enable tests - https://github.com/ecadlabs/taqueria/issues/377

File: tests/e2e/data/config-taquito-test-environment-invalid-config-network-url.json (hunk 5 above)
Comment (sinapsist) [reply tests/e2e/data/config-taquito-test-environment-invalid-config-network-url.json:57]: Added new issue to address it as feature request - https://github.com/ecadlabs/taqueria/issues/378

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 4 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:98]: Added issue to make this test - https://github.com/ecadlabs/taqueria/issues/379

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 7 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:42]: Fixed.

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 9 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:177]: Fixed. Added verification step that hashes are different

File: tests/e2e/taqueria-plugin-taquito.spec.ts (hunk 8 above)
Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:57]: We will do it on sandbox, not on the network - need to add additional tests there. On the network, we will check that the second contract does not exist by checking the table.

Comment (sinapsist) [reply tests/e2e/taqueria-plugin-taquito.spec.ts:59]: There is an issue with taquito-flextesa plugins tests -https://github.com/ecadlabs/taqueria/issues/405

**Review by alexzbusko — APPROVED** [review]
LGTM

**Review by mweichert — APPROVED** [review]
LGTM!
